### PR TITLE
[Feature] New swapbox design

### LINKED ISSUE #640: New swapbox redesign
- [ ] 3 tabs at the top (Have Ecobridge 1.5 enabled, Limit and Bridge Swap as disabled)
- [ ] Calculate the USD equivalent
- [ ] If the To estimate is under 2% slippage, show the percentage in grey, anything more than 2% show in Red next to the USD equivalent.
- [ ] Remove the current Max implementation, instead show the Balance and the value. The value is underlined, but both Balance and value should be clickable. Upon clicking, the Token portion should be filled with Maximum amount the wallet has.
- [ ] The circle in the center to be updated with new design, its a bit bigger now.

### PR DESCRIPTION
**Pull request by berteotti**
- Add 3 tabs at the top
- Calculate the USD equivalent and USD price impact
- Remove current Max implementation
- Show balance value below token input
- Update switch tokens button design
- Update swap button design to display selected trade platform design
- Add Swap and Send feature
- Add Swap Settings Buttons
- Closes #640, #645, #644 and #642

### CODE REVIEW

File: src/utils/computeFiatValuePriceImpact.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { CurrencyAmount, Percent, _10000, JSBI, currencyEquals, ZERO } from '@swapr/sdk'
+
+const ONE_HUNDRED_PERCENT = new Percent(_10000, _10000)
+
+export function computeFiatValuePriceImpact(
```
Comment (adamazad) [inline src/utils/computeFiatValuePriceImpact.ts:5]: Can we add some JSDoc descriptions?

File: src/utils/prices.ts [hunk 2]
```diff
@@ -97,6 +98,11 @@ export function warningSeverity(priceImpact: Percent | undefined): 0 | 1 | 2 | 3
   return 0
 }
 
+export function warningFiatSeverity(priceImpact: Percent | undefined): 0 | 3 {
+  if (!priceImpact?.lessThan(ALLOWED_FIAT_PRICE_IMPACT_HIGH)) return 3
+  return 0
+}
```
Comment (adamazad) [inline src/utils/prices.ts:104]: An enum can help with mapping severity names to numeric values.

File: src/hooks/useUSDValue.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  const { chainId } = useActiveWeb3React()
+  const stablecoin = chainId ? STABLECOIN_OUT[chainId] : undefined
+  const parsedAmountOut = tryParseAmount(AMOUNT_OUT, stablecoin, chainId)
+
+  const tradeExactOutAllPlatforms = useTradeExactOutAllPlatforms(currency, parsedAmountOut)
+
+  return useMemo(() => {
+    if (!currency || !chainId || !stablecoin || !tradeExactOutAllPlatforms) return undefined
+
+    if (currencyEquals(currency, stablecoin)) return new Price(currency, currency, '1', '1')
```
Comment (adamazad) [inline src/hooks/useUSDValue.ts:45]: This creates a technical debt - the list of stable coins continues to grow every day. We can remove it, Most DEX has stable coin pairs. On top of that, the peg between any stable coin and its fiat value is not a perfect 1:1.

File: src/hooks/useUSDValue.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    const allowHighTVLPlataforms = (trade: Trade | undefined) => {
+      if (!trade) return false
+      if (selectedTrade) return selectedTrade.platform.name === trade.platform.name
+      else return ALLOWED_PLATAFORMS[chainId].includes(trade.platform.name)
+    }
+
+    const calculateBestPrice = (trades: (Trade | undefined)[]): Price | undefined => {
```
Comment (adamazad) [inline src/hooks/useUSDValue.ts:53]: Out of `useEffect` and `useCallback`

**Review by adamazad — CHANGES_REQUESTED** [review]
I love this PR, The design feels refreshing! Great job!

**Comment by 0xVenky** [comment]
- [ ] We need to remove the USD value from all other places: Add Liquidity, Remove Liquidity, Claim rewards, Deposit and Stake
- [ ] Do not show USD value as zero :)

File: src/hooks/useUSDValue.ts (hunk 3 above)
Comment (berteotti) [reply src/hooks/useUSDValue.ts:45]: We can't simulate a trade for the same token so we are defining that: if the selected currency is the same as the one we defined as stable coin we do a 1:1 value match.

Comment (adamazad) [reply src/hooks/useUSDValue.ts:45]: I missed that. Makes sense now.

File: src/hooks/useUSDValue.ts (hunk 4 above)
Comment (berteotti) [reply src/hooks/useUSDValue.ts:53]: This functions are inside a `useMemo` and executed if one of this changes `[chainId, currency, selectedTrade, stablecoin, tradeExactOutAllPlatforms]`

File: src/utils/computeFiatValuePriceImpact.ts (hunk 1 above)
Comment (berteotti) [reply src/utils/computeFiatValuePriceImpact.ts:5]: Of course! Nice suggestion

**Comment by Mi-Lan** [comment]
Nice job @berteotti ! Looks really good! Will start listing some things I found.
- Seems that sushi part is not same as in figma
1) Seems that font-size for Sushi swap is 13px and on figma its 14.9px
2) On figma it says just "Sushi" but in pr it says "Sushiswap"
3) Icon seems a bit different then one of figma
This pr :
Screenshot 2022-03-07 at 13 04 59
Figma:
Screenshot 2022-03-07 at 13 05 51

**Comment by Mi-Lan** [comment]
Also Im not sure if Im doing something wrong but I cant see the option to add recipient when swapping (even though I have expert mode turned on) heres screenshot
Screenshot 2022-03-07 at 13 23 16

File: src/utils/computeFiatValuePriceImpact.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+ * @param {CurrencyAmount} fiatValueOutput
+ * @returns {Percent}
+ */
+export function computeFiatValuePriceImpact(
+  fiatValueInput: CurrencyAmount | undefined | null,
+  fiatValueOutput: CurrencyAmount | undefined | null
+): Percent | undefined {
+  if (!fiatValueOutput || !fiatValueInput) return undefined
+  if (!currencyEquals(fiatValueInput.currency, fiatValueOutput.currency)) return undefined
+  if (JSBI.equal(fiatValueInput.quotient, ZERO)) return undefined
```
Comment (Mi-Lan) [inline src/utils/computeFiatValuePriceImpact.ts:17]: Could this be consolidated into one statement with OR operators? like this 
`if ((!fiatValueOutput || !fiatValueInput) || !currencyEquals(fiatValueInput.currency, fiatValueOutput.currency) || JSBI.equal(fiatValueInput.quotient, ZERO))`

**Comment by Mi-Lan** [comment]
@0xVenky Also one more consideration if we are adding option for recipient for the expert mode should we update the text also in expert mode tooltip to include new feature 
Screenshot 2022-03-07 at 14 37 29

**Comment by berteotti** [comment]
> Also Im not sure if Im doing something wrong but I cant see the option to add recipient when swapping (even though I have expert mode turned on) heres screenshot Screenshot 2022-03-07 at 13 23 16

Can you try it again? It's working in my localhost and in the deployment link.

File: src/utils/prices.ts [hunk 6]
```diff
@@ -89,12 +90,23 @@ export function computeSlippageAdjustedAmounts(
   }
 }
 
+export const PRICE_IMPACT_NON_EXPERT = 4
+export const PRICE_IMPACT_HIGH = 3
+export const PRICE_IMPACT_MEDIUM = 2
+export const PRICE_IMPACT_LOW = 3
+export const NO_PRICE_IMPACT = 0
```
[the remark this answers is not in the document]
Comment (berteotti) [reply src/utils/prices.ts:97]: So, you are suggestion one object with the percentage value like `ALLOWED_PRICE_IMPACT_HIGH` and the result like `PRICE_IMPACT_HIGH`?

File: src/pages/Swap/Tabs.tsx [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+  &:disabled {
+    color: ${({ theme }) => theme.text6};
+    cursor: not-allowed;
+  }
+`
+
+const StyledEcoRouter = styled(EcoRouter)`
+  margin-right: 5px;
+`
```
[the remark this answers is not in the document]
Comment (berteotti) [reply src/pages/Swap/Tabs.tsx:68]: Sorry, what do you mean with "move styles above exported component"?

[the remark this answers is not in the document]
Comment (Mi-Lan) [reply src/pages/Swap/Tabs.tsx:68]: Im all files Styled Components are above the exported component.
This part should be at the bottom of Tabs.tsx
Screenshot 2022-03-07 at 18 34 11
And styles I highlighted should be on top....

File: src/utils/prices.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Mi-Lan) [reply src/utils/prices.ts:97]: Yeah something like that. Also Im looking now why does the price impact low and price impact high have the same value? Is that a typo? Should prime impact low be 1?

**Comment by Mi-Lan** [comment]
Ohh okay got the recipient in expert view to work...Not sure what that was!
- On that note seems that styling on figma and in pr don't match

- Figma:
Screenshot 2022-03-07 at 18 52 06

- This pr:
Screenshot 2022-03-07 at 18 52 30

File: src/utils/prices.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (berteotti) [reply src/utils/prices.ts:97]: Yeah, it's a typo. Going to fix it!

[the remark this answers is not in the document]
Comment (berteotti) [reply src/utils/prices.ts]: Fixed it, tell me if it's better now :)

File: src/utils/computeFiatValuePriceImpact.ts (hunk 5 above)
Comment (berteotti) [reply src/utils/computeFiatValuePriceImpact.ts]: Fixed!

File: src/pages/Swap/Tabs.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (berteotti) [reply src/pages/Swap/Tabs.tsx]: Fixed!

**Comment by berteotti** [comment]
> 1. Seems that font-size for Sushi swap is 13px and on figma its 14.9px
> 2. On figma it says just "Sushi" but in pr it says "Sushiswap"
> 3. Icon seems a bit different then one of figma


This is now fixed!

**Comment by niemam29** [comment]
Hey, i've founded that in some case when there is huge decimal value, the scaling is a bit broken. I think that it should be more nicely centred and the boxes should not interfere on each other
cache

**Comment by berteotti** [comment]
> Hey, i've founded that in some case when there is huge decimal value, the scaling is a bit broken. I think that it should be more nicely centred and the boxes should not interfere on each other cache

Hey @zamli what should we do in this case? Break into a new line, limit number's decimal places , add ellipsis?

**Comment by niemam29** [comment]
And next thing, not sure if it's a bug, but if user have set Slippage tolerance as 0.0 slippage box is greyed as it would be uneditable value
cache

**Comment by niemam29** [comment]
On lower screen resolutions than fhd (for ex. 1366x768), if user scroll down a bit it overlaps with "Routing through" inscription from the landing page.
cache

**Comment by adamazad** [comment]
I might be wrong, but shouldn't the USD value be shown regardless if I am yet to select a buy token?

image

**Comment by berteotti** [comment]
> I might be wrong, but shouldn't the USD value be shown regardless if I am yet to select a buy token?
> 
> image

I made the suggestion to remove the price before selecting a buy token to simplify the code and to always show the same price by selected platform (if there is no buy token the USD price was an average of chosen platforms). In this discussion @zamli pointed that we will always have a predefined buy token in the future, making this code obsolete. What do you think @adamazad ?

**Comment by Mi-Lan** [comment]
There seems to be some extra margin next to best price found as you can by the design it should be vertically aligned with Recipient
Design:
Screenshot 2022-03-21 at 13 07 17
This pr:
Screenshot 2022-03-21 at 13 05 34

**Review by denciu — APPROVED** [review]

**Comment by karczuRF** [comment]
> Hey, i've founded that in some case when there is huge decimal value, the scaling is a bit broken. I think that it should be more nicely centred and the boxes should not interfere on each other cache

For the mobile view the case which was pointed above is still not solved.
image



And btw should the lower box be not as width as the main one? 
image

**Comment by berteotti** [comment]
> And btw should the lower box be not as width as the main one? image

Regarding this, Swapbox has a new ecoRouter list design to be implemented.

**Comment by berteotti** [comment]
> There seems to be some extra margin next to best price found as you can by the design it should be vertically aligned with Recipient Design: Screenshot 2022-03-21 at 13 07 17 This pr: Screenshot 2022-03-21 at 13 05 34

This is now fixed!

**Comment by niemam29** [comment]
Problems with scaling still exists
cache

**Comment by berteotti** [comment]
> Problems with scaling still exists cache

Waiting for @zamli mobile design so this can be properly fixed.

**Review by 0xVenky — APPROVED** [review]

**Pull request merged**